rfcs/20181016-optimizer-unification.md

```diff
@@ -0,0 +1,398 @@
+```
+
+Current Keras signature:
+
+```Python
+Adagrad(lr=0.01, epsilon=None, decay=0.0)
+```
+
+Proposed signature:
+
+```Python
+Adagrad(learning_rate=0.001,
+        epsilon=1e-8,
+        decay=0.0,
+        initial_accumulator_value=0.1,
+        name="Adagrad")
+```
+
+### Adam
+
+Current TF signature:
+
+```Python
+AdamOptimizer(learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8, use_locking=False, name="Adam")
+```
+
+Current Keras signature:
+
+```Python
+Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)
+```
+
+Proposed signature:
+
+```Python
+Adam(learning_rate=0.001,
+     beta_1=0.9,
+     beta_2=0.999,
+     epsilon=1e-8,
+     decay=0.0,
+     amsgrad=False,
+     name="Adam")
+```
+
+### FTRL
+
+Current TF signature:
+
+```Python
+FtrlOptimizer(learning_rate,
+              learning_rate_power=-0.5,
+              initial_accumulator_value=0.1,
+              l1_regularization_strength=0.0,
+              l2_regularization_strength=0.0,
+              use_locking=False,
+              name="Ftrl",
+              accum_name=None,
+              linear_name=None,
+              l2_shrinkage_regularization_strength=0.0)
+```
+
+Proposed signature:
+
+```Python
+FTRL(learning_rate,
+     learning_rate_power=-0.5,
+     initial_accumulator_value=0.1,
+     l1_regularization_strength=0.0,
+     l2_regularization_strength=0.0,
+     name="FTRL",
+     l2_shrinkage_regularization_strength=0.0)
+```
+
+
+### RMSProp
+
+Current TF signature:
+
+```Python
+RMSPropOptimizer(learning_rate,
+                 decay=0.9,
+                 momentum=0.0,
+                 epsilon=1e-10,
+                 use_locking=False,
+                 centered=False,
+                 name="RMSProp")
+```
+
+Current Keras signature:
+
+```Python
+RMSprop(lr=0.001, rho=0.9, epsilon=None, decay=0.0)
+```
+
+Proposed signature:
+
+```Python
+RMSProp(learning_rate=0.001,
+        rho=0.9,
+        epsilon=1e-8,
+        decay=0.0,
+        centered=False,
+        name="RMSProp")
+```
+
+**Notes:**
+- The `rho` argument was named `decay` in TF. The `decay` argument is a standard argument on all adaptive learning-rate optimizers.
+
+
+### Adamax
+
+Current Keras signature:
+
+```Python
+Adamax(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0)
+```
+
+Proposed signature:
+
+```Python
+Adamax(learning_rate=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-8, decay=0.0, name="Adamax")
+```
+
+
+### Nadam
+
+Current Keras signature:
+
+```Python
+Nadam(lr=0.002, beta_1=0.9, beta_2=0.999, epsilon=None, schedule_decay=0.004)
+```
+
+Proposed signature:
+
+```Python
+Nadam(learning_rate=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-8, decay=0.0, name="Nadam")
+```
+
+
+
+---
+
+## Questions and Discussion Topics
+
+- Do you have a use case where you need to reuse an optimizer across different sets of weights? (note: this will still be doable with this proposal) Describe your use case.
```

Comment:
A common use case: each GPU has its own set of weights, and reuse the optimizer for every GPU, e.g.:
```python
for k in range(8):
    optimizer.apply_gradients(grads_and_vars[k])
```
Sure it's possible to create 8 optimizers, but then there will be 8 learning_rate to set.